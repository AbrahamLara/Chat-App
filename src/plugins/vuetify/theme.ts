import colors from 'vuetify/lib/util/colors';
import { ThemeOptions } from 'vuetify/types/services/theme';

const customThemeConfig: ThemeOptions = {
  themes: {
    dark: {
      accent: colors.blue.accent1,
    },
    light: {
      accent: colors.blue.accent1,
    },
  },
};

export default customThemeConfig;
