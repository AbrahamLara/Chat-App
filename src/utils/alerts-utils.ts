export enum ErrorPayloadType {
  REGISTER = 'register',
  LOGIN = 'login',
}

/**
 * An alert message object.
 */
export interface AlertMessage {
  /**
   * The message to display.
   */
  message: string;
}

/**
 * Describes an alert banner that displays messages for the user to see on a page.
 */
export interface AlertBarOptions extends AlertMessage {
  /**
   * Determines if the banner should be visible.
   */
  visible: boolean;

  /**
   * The type of alert.
   */
  type: string;
}

/**
 * Describes an error message for an invalid form field value.
 */
export interface FormAlertMessage extends AlertMessage {
  /**
   * The form field the error message is for. This may not to be provided if a form error wasn't due to an invalid field
   * value.
   */
  field?: string;
}

/**
 * Describes the payload to return for a failed request.
 */
export interface ErrorPayload {
  /**
   * The error type.
   *
   * Note: This might not actually be needed.
   */
  type: ErrorPayloadType;

  /**
   * The error messages produced. Right now the backend only serves form errors.
   */
  errors: FormAlertMessage[];
}

export const DEFAULT_ALERT_BAR_OPTIONS: AlertBarOptions = {
  visible: false,
  type: '',
  message: '',
};

/**
 * Determines if the give value is an error response object sent from a server request.
 */
export function isFormErrorMessage(response: any) {
  return typeof response === 'object' && response.type && response.errors;
}

/**
 * Creates a generic message object to display as an alert.
 */
export function createAlertMessage(message: string): AlertMessage {
  return { message };
}
