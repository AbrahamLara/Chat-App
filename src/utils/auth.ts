export interface RegexRule {
  regex: RegExp;

  rule: string;

  passed?: boolean;
}

/**
 * Password regex also used in the backend.
 * [PASSWORD_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L22-L41)
 */
export const PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*\-_]).{8,}$/;

/**
 * Email regex also used in the backend.
 * [EMAIL_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L43-L48)
 */
const EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export const EMAIL_FORM_RULES = [
  (email: string) => !!email || 'Email is required',
  (email: string) => EMAIL_REGEX.test(email) || 'Email must be valid.',
];

export const PASSWORD_FORM_RULES = [
  (password: string) => !!password || 'Password is required',
  (password: string) =>
    PASSWORD_REGEX.test(password) || 'Password must be valid.',
];

export const PASSWORD_REGEX_RULES: RegexRule[] = [
  { rule: '8 characters minimum', regex: /.{8,}/ },
  { rule: '1 uppercase character', regex: /[A-Z]/ },
  { rule: '1 lowercase character', regex: /[a-z]/ },
  { rule: '1 number', regex: /[0-9]/ },
  {
    rule: '1 special character: !, @, #, $, %, ^, &, *, -, _',
    regex: /[!@#$%^&*\-_]/,
  },
];
