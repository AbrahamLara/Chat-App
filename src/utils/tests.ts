/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 */
function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

export { resolveVuetifyAppDataWarning };
