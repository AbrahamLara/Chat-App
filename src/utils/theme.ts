// Themes as string enums since the app state theme should only be either of these string values.
export enum THEME {
  DARK = 'Dark',
  LIGHT = 'Light'
}

// The application theme background color hexes.
export const ThemeHexes = {
  [THEME.DARK]: '#121212',
  [THEME.LIGHT]: '#fff'
};
