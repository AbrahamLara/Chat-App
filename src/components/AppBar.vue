<template>
  <v-app-bar :class="{ 'pa-sai': true, white: !usingDarkMode }" app fixed>
    <router-link class="text-decoration-none" to="/" exact>
      <div
        :class="{
          'white--text text-h5 d-flex align-center': true,
          'grey--text text--darken-4': !usingDarkMode,
        }"
      >
        ChatApp
      </div>
    </router-link>
    <v-spacer></v-spacer>
    <div class="d-flex align-center">
      <router-link
        to="/auth/signin"
        :class="{
          'text-decoration-none text--accent': true,
          'text--secondary': $route.path !== '/auth/signin',
        }"
      >
        <div class="auth-btn text-center rounded pa-2 pl-2 pr-2" v-ripple>
          Sign In
        </div>
      </router-link>
      <router-link
        to="/auth/signup"
        :class="{
          'text-decoration-none text--accent': true,
          'text--secondary': $route.path !== '/auth/signup',
        }"
      >
        <div class="auth-btn text-center rounded pa-2 pl-2 pr-2" v-ripple>
          Sign Up
        </div>
      </router-link>
      <v-divider class="mx-2 my-auto" style="height: 25px" vertical></v-divider>
      <settings-btn></settings-btn>
    </div>
  </v-app-bar>
</template>

<style>
.auth-btn {
  width: 80px;
  user-select: none;
}
</style>

<script lang="ts">
import Vue from 'vue';
import Settings from '@/components/Settings.vue';
import { Component, PropSync } from 'vue-property-decorator';
import { THEME } from '@/utils/theme';

@Component({
  name: 'app-bar',
  components: { 'settings-btn': Settings },
})
export default class AppBar extends Vue {
  // Prop sync is used so that the app bar can handle determining if dark mode is in use by just receiving the theme as
  // it changes.
  @PropSync('theme', String) syncedTheme!: string;

  get usingDarkMode() {
    return this.syncedTheme === THEME.DARK;
  }

  mounted() {
    console.log(this.$route.path);
  }
}
</script>
