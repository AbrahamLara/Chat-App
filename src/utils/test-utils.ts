import VueRouter, { RouteConfig } from 'vue-router';
import Vuex from 'vuex';
import { AppState } from '@/store/store-states';
import appMutations from '@/store/mutations';
import appActions from '@/store/actions';
import appModules from '@/store/modules';
import { DEFAULT_ROOT_STATE } from '@/store';
import { FormAlertMessage } from '@/utils/alerts-utils';

/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * ```
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 * ```
 */
export function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

/**
 * Create a Vuex store instance for testing.
 *
 * TODO: Reuse to create store in main.ts.
 *
 * @param vx The Vuex value
 * @param customState A state object that modifies the default state
 */
export function createStore(
  vx: typeof Vuex,
  customState: Partial<AppState> = {}
) {
  const mutations = appMutations;
  const actions = appActions;
  const modules = appModules;

  return new vx.Store({
    state: { ...DEFAULT_ROOT_STATE, ...customState },
    actions,
    mutations,
    modules,
  });
}

/**
 * Creates a VueRouter instance for testing.
 *
 * TODO: Reuse to create router in main.ts.
 *
 * @param vr The VueRouter value
 * @param routes The route config. If not provided, then a default config is provided.
 */
export function createRouter(
  vr: typeof VueRouter,
  routes: RouteConfig[] = [{ path: '/' }]
) {
  return new vr({ routes, mode: 'history' });
}

/**
 * Creates a form error message for testing.
 */
export function createFormAlertMessage(
  field: string,
  message: string
): FormAlertMessage {
  return { field, message };
}
