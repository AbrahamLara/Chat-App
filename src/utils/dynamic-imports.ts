// This file contains dynamic imports of components to reduce bundle size when building project with webpack.

const getAbout = () =>
  import(/* webpackChunkName: "About" */ '@/components/About.vue');

const getAppBar = () =>
  import(/* webpackChunkName: "AppBar" */ '@/components/AppBar.vue');

const getAuthPage = () =>
  import(/* webpackChunkName: "AuthPage"  */ '@/views/AuthPage.vue');

const getFormCard = () =>
  import(/* webpackChunkName: "FormCard" */ '@/components/FormCard.vue');

const getNotFound = () =>
  import(/* webpackChunkName: "NotFound" */ '@/views/NotFound.vue');

const getRegexRulesTester = () =>
  import(
    /* webpackChunkName: "RegexRulesTester" */ '@/components/RegexRulesTester.vue'
  );

const getSettings = () =>
  import(/* webpackChunkName: "Settings" */ '@/components/Settings.vue');

const getSignIn = () =>
  import(/* webpackChunkName: "SignIn" */ '@/components/SignIn.vue');

const getSignUp = () =>
  import(/* webpackChunkName: "SignUp" */ '@/components/SignUp.vue');

const getSplash = () =>
  import(/* webpackChunkName: "Splash" */ '@/views/Splash.vue');

export {
  getAbout,
  getAppBar,
  getAuthPage,
  getFormCard,
  getNotFound,
  getRegexRulesTester,
  getSettings,
  getSignIn,
  getSignUp,
  getSplash,
};
