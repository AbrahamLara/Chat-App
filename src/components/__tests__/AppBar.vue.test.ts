import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';
import AppBar from '@/components/AppBar.vue';
import { createLocalVue, mount, ThisTypedMountOptions } from '@vue/test-utils';
import { THEME } from '@/utils/theme';

const localVue = createLocalVue();
Vue.use(VueRouter);
Vue.use(Vuetify);

describe('AppBar.vue.test.ts', () => {
  const mountFunction = function(options: ThisTypedMountOptions<any>) {
    const vuetify = new Vuetify();
    const router = new VueRouter();
    return mount(AppBar, {
      localVue,
      vuetify,
      router,
      ...options
    });
  };

  it('renders light mode correctly', () => {
    const wrapper = mountFunction({ propsData: { theme: THEME.LIGHT } });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders dark mode correctly', () => {
    const wrapper = mountFunction({ propsData: { theme: THEME.DARK } });
    expect(wrapper.element).toMatchSnapshot();
  });
});
