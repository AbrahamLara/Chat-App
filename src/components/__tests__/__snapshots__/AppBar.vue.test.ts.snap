// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar.vue.test.ts renders dark mode correctly 1`] = `
<header
  class="v-sheet theme--light elevation-1 v-toolbar v-app-bar v-app-bar--fixed pa-sai"
  style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
>
  <div
    class="v-toolbar__content"
    style="height: 64px;"
  >
    <a
      aria-current="page"
      class="text-decoration-none router-link-exact-active router-link-active"
      href="#/"
    >
      <div
        class="white--text text-h5 d-flex align-center"
      >
        
      ChatApp
    
      </div>
    </a>
     
    <div
      class="spacer"
    />
     
    <div
      class="d-flex align-center"
    >
      <a
        class="text-decoration-none"
        href="#/#login"
      >
        <div
          class="auth-btn text--secondary text-center rounded pa-2 pl-2 pr-2"
        >
          
        Login
      
        </div>
      </a>
       
      <a
        class="text-decoration-none"
        href="#/#signup"
      >
        <div
          class="auth-btn text--secondary text-center rounded pa-2 pl-2 pr-2"
        >
          
        Signup
      
        </div>
      </a>
       
      <hr
        aria-orientation="vertical"
        class="mx-2 my-auto v-divider v-divider--vertical theme--light"
        role="separator"
        style="height: 25px;"
      />
       
      <div
        class="v-menu"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light elevation-0 v-size--default"
          role="button"
          style="font-size: 16px;"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-cog-outline theme--light"
            />
          </span>
        </button>
        <!---->
      </div>
    </div>
  </div>
</header>
`;

exports[`AppBar.vue.test.ts renders light mode correctly 1`] = `
<header
  class="v-sheet theme--light elevation-1 v-toolbar v-app-bar v-app-bar--fixed pa-sai white"
  style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
>
  <div
    class="v-toolbar__content"
    style="height: 64px;"
  >
    <a
      aria-current="page"
      class="text-decoration-none router-link-exact-active router-link-active"
      href="#/"
    >
      <div
        class="white--text text-h5 d-flex align-center grey--text text--darken-4"
      >
        
      ChatApp
    
      </div>
    </a>
     
    <div
      class="spacer"
    />
     
    <div
      class="d-flex align-center"
    >
      <a
        class="text-decoration-none"
        href="#/#login"
      >
        <div
          class="auth-btn text--secondary text-center rounded pa-2 pl-2 pr-2"
        >
          
        Login
      
        </div>
      </a>
       
      <a
        class="text-decoration-none"
        href="#/#signup"
      >
        <div
          class="auth-btn text--secondary text-center rounded pa-2 pl-2 pr-2"
        >
          
        Signup
      
        </div>
      </a>
       
      <hr
        aria-orientation="vertical"
        class="mx-2 my-auto v-divider v-divider--vertical theme--light"
        role="separator"
        style="height: 25px;"
      />
       
      <div
        class="v-menu"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light elevation-0 v-size--default"
          role="button"
          style="font-size: 16px;"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-cog-outline theme--light"
            />
          </span>
        </button>
        <!---->
      </div>
    </div>
  </div>
</header>
`;
